File: lab5c++/teacher.cpp
#include "teacher.h"


float teacher::countPayment() const {
    float res =0;
    for(int i=0; i<subjectsHours.size(); ++i)
        res+=subjectsHours[i]*paymentForEachHour;
    return res;
}

ostream &operator<<(ostream &os, const teacher &teacher) {

    string name = teacher.getFullname();
    tm date = teacher.getBirth();

    char *dateString = new char[100];
    strftime(dateString, 100, "date of birth: %x", &date);

    os <<name<<endl<< dateString<<endl<<"age: "<<teacher.getAge()<<endl;
    delete[] dateString;
    os<<"subjects: ";
    for(int i =0; i < teacher.subjects.size();++i)
        os<<teacher.subjects[i]<<", "<<teacher.subjectsHours[i]<<"; ";
    os<<"\npayment: "<<teacher.countPayment()<<endl;

    return os;
}

int findBiggestPaymentAge(teacher* a, int m) {
    float biggestPayment =0;
    int age =0;
    for(int i=0;i<m;++i)
    {
        float tmp = a[i].countPayment();
        if(tmp - biggestPayment > 0)
        {
            biggestPayment = tmp;
            age = a[i].getAge();
        }

    }
    return age;
}

teacher::teacher(): person() {
    bool flag = true;
    while(flag){
        cout<<"Enter subject:";
        string s;
        cin>>s;
        cout<<"Enter hours:";
        int n;
        cin>>n;
        subjects.push_back(s);
        subjectsHours.push_back(n);
        cout<<"Add another subject(y/n)?";
        string ans;
        cin>>ans;
        if(ans!="y")
            flag=false;
    }
}

File: lab5python/teacher.py
from person import *


class Teacher(Person):
    def __init__(self):
        super(__class__, self).__init__()
        self.paymentForEachHour = 250
        self.subjects = []
        self.subjectsHours = []
        flag = True
        while flag:
            a=input("Enter subject: ")
            b=int(input("Enter hours: "))
            self.subjects.append(a)
            self.subjectsHours.append(b)
            ans=input("Add another subject(y/n)?")
            if(ans != "y"):
                flag=False

    def countPayment(self):
        res=0
        for i in self.subjectsHours:
            res+=self.paymentForEachHour*i
        return res

    def display(self):
        print(self.fullName)
        date = self.birth.strftime("%d/%m/%Y")
        print("date of birth: " + date)
        print("age: " + str(self.getAge()))
        subjStr=''
        for i in range(0, len(self.subjects)):
            subjStr+="{}, {}; ".format(self.subjects[i],
                                       self.subjectsHours[i])
        print("subjects: " + subjStr)
        print("payment: " + str(self.countPayment()))

def findBiggestPaymentAge(arr):
    biggestPayment = 0
    age = 0
    for a in arr:
        tmp = a.countPayment()
        if(tmp - biggestPayment > 0):
            biggestPayment=tmp
            age= a.getAge()
    return age